packages/foundry/contracts/hooks/OrderHook.sol

```diff
@@ -0,0 +1,476 @@
+
+        IERC20(order.tokenIn).approve(address(i_permit2), 0);
+        IERC20(order.tokenOut).safeTransfer(order.trader, amountOut);
+        order.orderStatus = OrderStatus.EXECUTED;
+    }
+
+    /// @notice Checks if the order should be executed
+    /// @dev If the order is a stop loss and the tokenIn price is less than or equal to the trigger price, the order should be executed
+    /// @dev If the order is a buy stop and the tokenIn price is greater than or equal to the trigger price, the order should be executed
+    /// @dev If the order is a buy limit and the tokenIn price is less than or equal to the trigger price, the order should be executed
+    /// @dev If the order is a take profit and the tokenIn price is greater than or equal to the trigger price, the order should be executed
+    /// @param order The order to check
+    /// @param tokenInlPrice The price of the tokenIn in the pool where order is placed
+    function _shouldExecuteOrder(
+        Order storage order,
+        uint256 tokenInlPrice
+    ) internal view returns (bool) {
+        if (
+            order.orderType == OrderType.STOP_LOSS &&
+            tokenInlPrice <= order.triggerPrice
+        ) {
+            return true;
+        } else if (
+            order.orderType == OrderType.BUY_STOP &&
+            tokenInlPrice >= order.triggerPrice
+        ) {
+            return true;
+        } else if (
+            order.orderType == OrderType.BUY_LIMIT &&
+            tokenInlPrice <= order.triggerPrice
+        ) {
+            return true;
+        } else if (
+            order.orderType == OrderType.TAKE_PROFIT &&
+            tokenInlPrice >= order.triggerPrice
+        ) {
+            return true;
+        }
+        return false;
+    }
+
+    /// @notice Calculate the limit of the tokenOut for a swap
+    /// @dev The limit is calculated based on the slippage tolerance
+    /// @param amountIn Amount of tokenIn
+    /// @param tokenInPrice Current price of tokenIn
+    /// @param tokenOutPrice Current price of tokenOut
+    /// @param slippageTolerance Slippage tolerance
+    /// @return limit The minimum amount of tokenOut to receive
+    function _calculateLimit(
+        uint256 amountIn,
+        uint256 tokenInPrice,
+        uint256 tokenOutPrice,
+        uint8 slippageTolerance
+    ) internal pure returns (uint256 limit) {
+        uint256 estimatedAmountOut = (amountIn * tokenOutPrice) / tokenInPrice;
+        limit = ((estimatedAmountOut * (100 - slippageTolerance)) / 100);
+    }
+
+    /*////////////////////////////////////////////////
+    ///////   PUBLIC, VIEW, GETTER FUNCTIONS   ///////
+    ////////////////////////////////////////////////*/
+
+    /// @notice Get the current prices of the tokens in a pool
+    /// @param pool The pool address
+    function getPoolPrice(
+        address pool
+    ) public view returns (uint256[] memory balancesLiveScaled18) {
+        balancesLiveScaled18 = i_vault.getCurrentLiveBalances(pool);
```

Comment:
Here again we're getting the balances and using them as prices directly to check whether we should execute the orders. Could you explain the logic behind it?

Reply:
Thank you for the question. My approach here uses token balances as a proxy for prices due to how Balancer pools maintain price ratios relative to token quantities in the pool.

The logic is based on the idea that a change in the relative token balances (especially scaled to 18 decimals for uniformity) reflects price shifts, which can act as a trigger for order execution offchain.

Like after a swap balances change. So, we use the ratio of new balances as price to evaluate existing orders and notify users offchain if their orders are executable.

I think the problem is in the `executeOrder()` function, where we use the balance of `tokenIn` in `_shouldExecuteOrder` and compare it with `triggerPrice`. We need to send the ratio of `tokenIn` to `tokenOut` and then compare with `triggerPrice` according to current implementation.

However, this approach doesn’t account for the pool’s invariant directly, which could impact accuracy.

Reply:
We can use an oracle instead if that works out best for the balancer protocol